File: 엄마호랑이반/8주차 문제/박정환/보물섬.py
from collections import deque

n, m = map(int, input().split())
board = [list(input()) for _ in range(n)]

# print(board)

dx = [-1, 0, 1, 0]
dy = [0, 1, 0 , -1]

def bfs(i,j):
  q = deque()
  q.append((i,j))
  visited[i][j]=1
  cnt=0
  while q:
    x, y = q.popleft()
    for i in range(4):
      xx = x+dx[i]
      yy = y+dy[i]
      if 0<= xx < n and 0<= yy < m and board[xx][yy]=='L' and visited[xx][yy]==0:
        visited[xx][yy]=visited[x][y]+1
        cnt = max(cnt, visited[xx][yy])
        q.append((xx,yy))
  return cnt-1

result=0
for i in range(n):
  for j in range(m):
    visited = [[0]*m for _ in range(n)]
    if board[i][j] == 'L':
      result = max(result, bfs(i,j))

print(result)

File: 엄마호랑이반/8주차 문제/이다영/보물섬.java
package temp_dev;

import java.util.*;
import java.io.*;

public class TestClass {
	
	public static int a;
	public static int b;
	public static String[][] island;
	public static int result = 0; 
	
	public static void main(String[] args) throws Exception {
		 BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		 
		 String[] ab = br.readLine().split(" ");
		 a = Integer.parseInt(ab[0]);
		 b = Integer.parseInt(ab[1]);
		 island = new String[a][b];
		 
		 for (int i = 0; i < a; i++) {
			 island[i] = br.readLine().split("");
		 }
		 
		 for (int x = 0; x < a; x++) {
			 for (int y = 0; y < b; y++) {
				 if(!"W".equals(island[x][y])) {
					 bfs(x, y);
				 }
			 }
		 }
		 
		 System.out.println(result);
	}
	
	public static void bfs(int sx, int sy) {
		int[] dx = new int[] {0, 1, 0, -1};
		int[] dy = new int[] {-1, 0, 1, 0};
		int[][] dis = new int[a][b];
		for(int i = 0; i < a; i++) {
			Arrays.fill(dis[i], -1);
		}
		dis[sx][sy] = 0;
		
		Deque<int[]> dq = new ArrayDeque<>();
		dq.addLast(new int[]{sx, sy});
		
		int cnt = 0;
		while (!dq.isEmpty()) {
			int[] tmp = dq.pollFirst();
			int x = tmp[0]; 
			int y = tmp[1];
			
			for (int z = 0; z < 4; z++) {
				int xx = x + dx[z];
				int yy = y + dy[z];
				
				if ( 0 <= xx && xx < a && 0 <= yy && yy < b && 
					"L".equals(island[xx][yy]) && 
					dis[xx][yy] == -1) {
					
					dis[xx][yy] = dis[x][y] + 1; 
					dq.addLast(new int[] {xx, yy});
					cnt = dis[xx][yy];
				}
			}
			
			result = Math.max(cnt , result);
		}
	}
}
